src/representation_theory/basic.lean

```diff
@@ -0,0 +1,116 @@
+/-
+Copyright (c) 2022 Antoine Labelle. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Antoine Labelle
+-/
+import algebra.module.basic
+import data.equiv.module
+import algebra.module.linear_map
+import algebra.monoid_algebra.basic
+import linear_algebra.trace
+
+/-!
+# Monoid representations
+
+This file introduces monoid representations and their characters and proves basic lemmas about them,
+including equivalences between different definitions of representations.
+
+## Main definitions
+
+  * `as_module`
+  * `rep_space`
+  * `character`
+
+## Implementation notes
+
+A representation of a monoid `G` over a commutative semiring `k` is implemented as a `k`-module `V`
+together with a `distrib_mul_action G V` instance and a `smul_comm_class G k V` instance.
+
+Alternatively, one can use a monoid homomorphism `G →* (V →ₗ[k] V)`. The definitions `as_monoid_hom`
+and `rep_space` allow to go back and forth between these two definitions.
+-/
+
+open monoid_algebra
+
+namespace representation
+
+section
+variables (k G V : Type*) [comm_semiring k] [monoid G] [add_comm_monoid V]
+variables [module k V] [distrib_mul_action G V] [smul_comm_class G k V]
+
+/--
+A k-linear representation of G on V can be thought of as
+an algebra map from `monoid_algebra k G` into the k-linear endomorphisms of V.
+-/
+noncomputable def as_algebra_hom : monoid_algebra k G →ₐ[k] (module.End k V) :=
+  (lift k G _) (distrib_mul_action.to_module_End k V)
+
+lemma as_algebra_hom_def :
+  as_algebra_hom k G V = (lift k G _) (distrib_mul_action.to_module_End k V) := rfl
+
+@[simp]
+lemma as_algebra_hom_single (g : G):
+  (as_algebra_hom k G V (finsupp.single g 1)) = (distrib_mul_action.to_module_End k V) g :=
+by simp [as_algebra_hom_def]
+
+/--
+A k-linear representation of G on V can be thought of as
+a module over `monoid_algebra k G`.
+-/
+noncomputable instance as_module : module (monoid_algebra k G) V :=
```

Comment:
I think this should be a `def`. I didn't check in detail, but I think this will create non-defeq loops in the typeclass hierarchy if you start with a `module (monoid_algebra k G) V`, and deduce from there the instances on line 39.

cc @eric-wieser

Reply:
I don't think the loop @jcommelin is describing is something to worry about; it's the other side of the loop that @jcommelin  is describing that is dangerous, not this side, as instances should construct data for bigger types out of smaller types.

The danger here is whether this conflicts with `semiring.to_module`. It's hard for me to check that right now as you're over 1000 commits behind master; I pushed a merge commit.

The test would look something roughly like:
```lean
instance : smul_comm_class G k (monoid_algebra k G) := sorry

example : representation.as_module _ _ _ = (semiring.to_module : module (monoid_algebra k G) (monoid_algebra k G)) :=
begin
  -- my attempt at a proof before realizing I wasn't on master
  -- if this is false, then we have a problem
  -- if this isn't provable by refl, then we maybe have a problem.
  ext,
  dsimp [monoid_algebra.mul_apply, (•), as_algebra_hom_def, lift_apply],
  rw [monoid_algebra.mul_apply, lift_apply],
  simp only [linear_map.finsupp_sum_apply, finsupp.sum_apply],
  dsimp,
  congr' 1 with i ai,
  simp_rw ←finsupp.single_apply,
  sorry
end
```

Reply:
@eric-wieser I'm sorry I don't think I understand exactly the subtilities of the potential dangers here. I'm not sure what your test is supposed to do. Have you been able to check if there are possible conflicts or not?